bundles/org.openhab.binding.dirigera/src/main/java/org/openhab/binding/dirigera/internal/network/Websocket.java

```diff
@@ -0,0 +1,251 @@
+        }
+    }
+
+    public boolean isRunning() {
+        return websocketClient.isPresent() && session.isPresent() && session.get().isOpen();
+    }
+
+    public void stop() {
+        increase(STOPS);
+        internalStop();
+    }
+
+    private void internalStop() {
+        session.ifPresent(session -> {
+            session.close();
+        });
+        websocketClient.ifPresent(client -> {
+            try {
+                client.stop();
+                client.destroy();
+            } catch (Exception e) {
+                logger.warn("DIRIGERA WS exception stopping running client");
+            }
+        });
+        websocketClient = Optional.empty();
+        this.session = Optional.empty();
+    }
+
+    public void dispose() {
+        internalStop();
+        disposed = true;
+    }
+
+    public void ping() {
+        session.ifPresentOrElse((session) -> {
+            try {
+                // build ping message
+                String pingId = UUID.randomUUID().toString();
+                pingPongMap.put(pingId, Instant.now());
+                session.getRemote().sendPing(ByteBuffer.wrap(pingId.getBytes()));
+                increase(PINGS);
+            } catch (IOException e) {
+                logger.info("DIRIGERA WS ping failed with exception {}", e.getMessage());
+            }
+        }, () -> {
+            logger.info("DIRIGERA WS ping found no session - restart websocket");
+        });
+    }
+
+    /**
+     * endpoints
+     */
+
+    @OnWebSocketMessage
+    public void onTextMessage(String message) {
+        increase(MESSAGES);
+        gateway.websocketUpdate(message);
+    }
+
+    @OnWebSocketFrame
+    public void onFrame(Frame frame) {
+        if (Frame.Type.PONG.equals(frame.getType())) {
+            ByteBuffer buffer = frame.getPayload();
+            byte[] bytes = new byte[frame.getPayloadLength()];
+            for (int i = 0; i < frame.getPayloadLength(); i++) {
+                bytes[i] = buffer.get(i);
+            }
+            String paylodString = new String(bytes);
+            Instant sent = pingPongMap.remove(paylodString);
+            if (sent != null) {
+                long durationMS = Duration.between(sent, Instant.now()).toMillis();
+                statistics.put(PING_LATENCY, durationMS);
+                statistics.put(PING_LAST, Instant.now());
+            } else {
+                logger.info("DIRIGERA WS receiced pong without ping {}", paylodString);
+            }
+        } else if (Frame.Type.PING.equals(frame.getType())) {
+            session.ifPresentOrElse((session) -> {
+                logger.trace("DIRIGERA onPing ");
+                ByteBuffer buffer = frame.getPayload();
+                try {
+                    session.getRemote().sendPong(buffer);
+                } catch (IOException e) {
+                    logger.info("DIRIGERA WS onPing answer exception {}", e.getMessage());
```

Comment:
Please check the log level everywhere:
For reference: https://www.openhab.org/docs/developer/guidelines.html#f-logging

Reply:
Mostly all exceptions are covered with `warn` level. 
Adapted some but in general during "normal operation" there are no trace outputs

9 trace - some stats
18 debug - mainly token & websocket
42 info - majority covered by customDebug flag
2 errors if API cannot be created

Reply:
See review round specific to logging. Will continue from there.